Add parsing method for Matrix Stats

tlrx opened the pull request:
This PR adds the parsing logic for the `InternalMatrixStats` aggregation.

tlrx commented:
@cbuescher @javanna This is ready to be reviewed.

tlrx commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/InternalMatrixStats.java:
```diff
@@ -157,7 +157,7 @@ MatrixStatsResults getResults() {
     @Override
     public XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
-        if (results != null && results.getFieldCounts().keySet().isEmpty() == false) {
+        if (results != null) {
```
I had to change this so that at parsing time we can make the distinction between no existing results and empty results. 

When there is no results, the rendered output will be something like `"matrix_stats":{}` and empty results will be  `"matrix_stats":{"fields":[]}`. This allows us to reproduce the same behavior of MatrixStats methods in the transport client and parsed implementations.

I'm not sure that "no results" is a real use case but InternalMatrixStats has specific checks at several for `results` being null (and it's also randomized in tests) so it let me think that it can happen.

cbuescher replied:
Looking at `InternalMatrixStats#doReduce()` it seems to me that at least after the aggregations are reduced (which I think should be always the case for aggregations that are rendered to REST) the `results` should always be set. Maybe we can check with somebody who knows more about the internals and work with that assumption on the client side instead of making this change.

tlrx replied:
I agree, after looking again `results` is never null once the aggregations are reduced. It can though be `null` at the transport level when shard results are sent over the wire before being reduced on the coordinating node.

It explains why `results` is randomly set to null in `InternalMatrixStats` because serialization is tested there. I reverted this change and only test the normal case.

cbuescher commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/ParsedMatrixStats.java:
```diff
@@ -0,0 +1,247 @@
+
+    @Override
+    public long getDocCount() {
+        throw new UnsupportedOperationException();
```
If we don't render this to the REST response, but its part of the MatrixStats interface, should we either add this value to the output or remove the method from the interface?

tlrx replied:
I think it makes sense to add this field to the REST response. I can do that in another PR against master.

javanna replied:
++

cbuescher commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/ParsedMatrixStats.java:
```diff
@@ -0,0 +1,247 @@
+        if (covariances == null) {
+            return Double.NaN;
+        }
+        return checkedGet(checkedGet(covariances, fieldX), fieldY);
```
nit: I'm not entirely sure this does exactly what getValFromUpperTriangularMatrix() does (e.g. I think swapping the fieldX, fieldY arguments should return the same result in the original implementation, I'm not entirely sure though). It should be possible to declare that method as static in `MatrixStatsResults` and then use it from here to get the same behaviour.

tlrx replied:
Good idea, thanks

cbuescher commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/ParsedMatrixStats.java:
```diff
@@ -0,0 +1,247 @@
+        if (fieldX != null && fieldX.equals(fieldY)) {
+            return 1.0;
+        }
+        return checkedGet(checkedGet(correlations, fieldX), fieldY);
```
Same here about using getValFromUpperTriangularMatrix() if possible.

cbuescher commented on modules/aggs-matrix-stats/src/test/java/org/elasticsearch/search/aggregations/matrix/stats/InternalMatrixStatsTests.java:
```diff
@@ -18,35 +18,62 @@
-                                                     Map<String, Object> metaData) {
+    public void setUp() throws Exception {
+        super.setUp();
+        hasMatrixStatsResults = randomBoolean();
```
Based on the discussion above (whether we need to account for the "has no results" case at all after "reduce", I would opt for at least testing it less frequent, and thest the "result" case mostly.

tlrx replied:
Sure

cbuescher commented on modules/aggs-matrix-stats/src/test/java/org/elasticsearch/search/aggregations/matrix/stats/InternalMatrixStatsTests.java:
```diff
@@ -100,4 +127,55 @@ public void testReduceRandom() {
+        }
+
+        final String unknownField = randomAlphaOfLength(3);
+        final String fieldX = randomFrom(unknownField, randomAlphaOfLength(3));
```
I'm not sure I understand what this is doing?

tlrx replied:
Yeah, it's not really readable. I changed that. The idea is to test getCovariance/getCorrelation with various unknown fields.

javanna commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/ParsedMatrixStats.java:
```diff
@@ -0,0 +1,247 @@
+            matrixStats.skewness = new LinkedHashMap<>(size);
+            matrixStats.kurtosis = new LinkedHashMap<>(size);
+            matrixStats.covariances = new LinkedHashMap<>(size);
+            matrixStats.correlations = new LinkedHashMap<>(size);
```
do we need to maintain insertion order in all these maps?

tlrx replied:
No, but at least on `counts` so that we can render the parsed aggregation with the same order of fields stats.

tlrx commented:
Thanks @cbuescher and @javanna for your reviews. I updated the code according to your comments.

I think this PR could be merged without elastic/elasticsearch#2477. Just in case I added a //norelease comment to not forget about updating the parsing logic and test once the change in core is merged.

cbuescher commented on modules/aggs-matrix-stats/src/test/java/org/elasticsearch/search/aggregations/matrix/stats/InternalMatrixStatsTests.java:
```diff
@@ -74,7 +74,7 @@ protected InternalMatrixStats createTestInstance(String name, List<PipelineAggre
         runningStats.add(fields, values);
         MatrixStatsResults matrixStatsResults = hasMatrixStatsResults ? new MatrixStatsResults(runningStats) : null;
-        return new InternalMatrixStats("_name", 1L, runningStats, matrixStatsResults, Collections.emptyList(), Collections.emptyMap());
+        return new InternalMatrixStats(name, 1L, runningStats, matrixStatsResults, Collections.emptyList(), Collections.emptyMap());
```
Good catch ;-)

tlrx commented:
Thanks @javanna and @cbuescher
